### ISSUE TITLE
Ktor IO dependency should be an `api` dependency

### ISSUE THREAD

**Issue by Jeffset** [kgit2/kommand#1]
As far as I see, ktor-io should be an `api` dependency rather than `implementation`, as it is present in library's public API, e.g. in `Writer` or `Reader` classes. So, if anybody wants to just use the library without knowing anything about ktor and stuff, they'll be in a bit of trouble, as they'll end up with, plainly speaking, broken compilation classpath.

Another way to go would be to fully encapsulate IO API, leaving ktor an implementation detail indeed. This way is preferable, imo, though requires more work.

**Comment by BppleMan** [comment]
You're right, but I'm considering whether to use okio, which is a standalone io library that might be easier for users to learn.  If I'm not going to expose any 3rd party io, maybe I really need fully encapsulated io.  Do you have more views on this?

**Comment by Jeffset** [comment]
I think, fully incapsulating IO is a great idea, go with it if you can. This way you can decide on what actual IO backend to use and update without introducing breaking changes for the clients.

**Comment by BppleMan** [comment]
I have been thinking about this issue recently. Now kotlin-multiplatform does not have a standard io protocol or implementation. Although ktor-io belongs to the first-party library, it is not in the stdlib, which creates a problem. If every third-party library Providing your own set of io standards is very unfriendly to users. Every time users use a library, they need to learn the io api corresponding to the library. This is really ugly, so I considered using ktor-io at the beginning. But before kotlin officially proposes any standard library implementation, I don't think more io implementations should be added, which is meaningless work

**Comment by BppleMan** [comment]
Anyone can discuss below

**Comment by sanyavertolet** [comment]
In our projects we use [square/okio](https://github.com/square/okio) which is nice (has 8k stars). At least, it seems more popular, so it might be better to use it, nevertheless, I'm not sure it fits you

**Comment by BppleMan** [comment]
> In our projects we use [square/okio](https://github.com/square/okio) which is nice (has 8k stars). At least, it seems more popular, so it might be better to use it, nevertheless, I'm not sure it fits you

Using okio I think is acceptable, But more votes are needed

**Comment by sanyavertolet** [comment]
Any updates here?

**Comment by BppleMan** [comment]
Now, we can migrate to kotlinx-io
https://youtrack.jetbrains.com/issue/KTOR-6030/Migrate-to-new-kotlinx.io-library

**Comment by BppleMan** [comment]
In fact, the current approach is a simple io encapsulation, and the exposed APIs are `readLine` and `lines`.

https://github.com/kgit2/kommand/blob/f26c0615969f3d3047fbb0ce3c3d83db04f29900/src/commonMain/kotlin/com/kgit2/io/Reader.kt#L12-L37

ktor-io only acts as a middle layer. It neither makes real system calls nor uses its external API. It only acts as a buffer to assist the above API.

Therefore no changes will be made for now, the api is simple enough to provide full functionality
